text/0053-enum-entities.md

```diff
@@ -0,0 +1,146 @@
+# Enumerated Entity Types
+
+## Timeline
+
+- Started: 2024-02-20
+
+## Summary
+
+Extend schemas to support declared enumerations of entity-typed values, analogous to how schemas can currently be used to enumerate a finite list of `Action`-typed values.
+
+## Basic example
+
+An enumerated entity type is declared as a normal entity type, but includes the keyword `enum` followed by the list of legal entity UIDs. Here is a simple example:
+```
+entity User;
+entity Color enum ["red", "blue", "green"];
+entity Task {
+    owner: User,
+    name: String,
+    status: Color
+};
+action UpdateTask
+    appliesTo { principal: [User], resource: [Task] };
+```
+These data definitions could be used in policies such as 
+```
+permit(
+    principal,
+    action == Action::"UpdateTask",
+    resource)
+when {
+    principal == resource.owner &&
+    resource.status != Color::"red"
+};
+```
+
+## Motivation
+
+Enumerated types are useful when you have a fixed set of possible values, and the only thing you want to do with the values is compare them for equality. While you could effectively treat an entity type as an enumeration now, without declaring it in a schema, you gain some benefits by declaring it:
+
+- The validator can error on uses of illegal enumerated values, e.g., flagging the typo `resource.status != Color::"Red"` in the `when` clause in the basic example. 
+- When using a policy analyzer, it can always generate request and entity store instances where the enumerated entity type has declared-valid values, rather than random UIDs.
+- When using an IDE or web-based policy builder, the enumeration can inform auto-completion suggestions. For the basic example above, in an auto-completing IDE writing `resource.status != ` ... would pop up the three options.
+
+## Detailed design
+
+An enumerated entity `Foo` is declared by writing
+```
+entity Foo enum [ … ];
+```
+where `[` … `]` is the list of allowed values. These values can be strings or identifiers.
```

Comment:
I don't think I like throwing in the "or identifiers" on the end.  Also, none of the existing examples in this writeup take advantage of the "or identifiers" part: in all of the examples in the RFC, the values are strings, not identifiers. 

Here's an example of what the "or identifiers" allows, but I think we should disallow:
```
entity Color enum [red, blue, green];
```
vs how it appears in all the examples currently:
```
entity Color enum ["red", "blue", "green"];
```
This latter form seems more intuitive, since you'll later have to reference the values as `Color::"red"`, `Color::"blue"`, etc.

Reply:
That's fair. We can drop the ID part and potentially add it later if there's interest in that.